File: scripts/_lib/http.py
"""
Shared HTTP / log / atomic-write helpers for Python sync scripts.

- ``fetch_with_retry(url, headers=...)`` — 3 attempts, exponential backoff
  (200/400/800 ms) with ±25% jitter. Retries on network errors and on
  HTTP 429/5xx. Returns the final ``urllib.response`` body (str) on
  success, ``None`` on terminal failure.
- ``RateLimiter(rps)`` — token bucket. ``rl.take()`` blocks until the
  next slot is free.
- ``write_json_atomic(path, data)`` — writes to ``<path>.tmp``, fsyncs,
  then ``os.replace`` (atomic on POSIX, same filesystem).
- ``json_log(**fields)`` — structured log line to stdout.
"""
from __future__ import annotations

import json
import os
import random
import sys
import time
import urllib.error
import urllib.request
from contextlib import contextmanager
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterator

DEFAULT_DELAYS_MS = (200, 400, 800)


def _jitter(base_ms: int) -> float:
    spread = base_ms * 0.25
    return base_ms + random.uniform(-spread, spread)


def fetch_with_retry(
    url: str,
    *,
    headers: dict[str, str] | None = None,
    timeout: int = 15,
    delays_ms: tuple[int, ...] = DEFAULT_DELAYS_MS,
) -> str | None:
    """Returns response body as str on success, None on terminal failure."""
    attempts = len(delays_ms) + 1
    for i in range(attempts):
        try:
            req = urllib.request.Request(url, headers=headers or {})
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                if 200 <= resp.status < 300:
                    return resp.read().decode("utf-8", errors="replace")
                if resp.status < 500 and resp.status != 429:
                    return None  # 4xx not worth retrying
        except urllib.error.HTTPError as exc:
            if exc.code < 500 and exc.code != 429:
                return None
        except (urllib.error.URLError, TimeoutError, ConnectionError):
            pass
        if i < len(delays_ms):
            time.sleep(_jitter(delays_ms[i]) / 1000.0)
    return None


class RateLimiter:
    def __init__(self, rps: float) -> None:
        self.interval = 1.0 / max(0.1, rps)
        self.next = 0.0

    def take(self) -> None:
        now = time.monotonic()
        wait = max(0.0, self.next - now)
        self.next = max(now, self.next) + self.interval
        if wait > 0:
            time.sleep(wait)


def write_json_atomic(path: Path, data: Any) -> None:
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    with tmp.open("w", encoding="utf-8") as fh:
        json.dump(data, fh, indent=2)
        fh.write("\n")
        fh.flush()
        os.fsync(fh.fileno())
    os.replace(tmp, path)


def json_log(**fields: Any) -> None:
    payload = {"ts": datetime.now(timezone.utc).isoformat(), **fields}
    if isinstance(payload.get("error"), BaseException):
        payload["error"] = repr(payload["error"])
    sys.stdout.write(json.dumps(payload) + "\n")
    sys.stdout.flush()


@contextmanager
def timed(operation: str) -> Iterator[dict[str, Any]]:
    """Context manager that logs duration + counters when exiting."""
    start = time.monotonic()
    counters: dict[str, Any] = {"recordsProcessed": 0, "ok": 0, "failed": 0}
    try:
        yield counters
        json_log(
            operation=operation,
            durationMs=int((time.monotonic() - start) * 1000),
            **counters,
        )
    except BaseException as exc:  # noqa: BLE001
        json_log(
            operation=operation,
            durationMs=int((time.monotonic() - start) * 1000),
            error=exc,
            **counters,
        )
        raise

File: scripts/_lib/http.mjs
/**
 * Shared HTTP / log / atomic-write helpers for sync scripts.
 *
 * - `fetchWithRetry(url, init, opts)` — 3 attempts, exponential backoff
 *   (200/400/800ms) with ±25% jitter. Retries on network errors and on
 *   HTTP 429/5xx. Returns the final Response (caller checks `.ok`).
 * - `RateLimiter(rps)` — simple token-bucket. `await rl.take()` before
 *   each external call.
 * - `writeJsonAtomic(path, data)` — writes to `<path>.tmp`, fsyncs, then
 *   renames over `path`. Atomic on POSIX (same filesystem).
 * - `jsonLog(payload)` — structured log line to stdout. All fields:
 *   `{ts, operation, durationMs, recordsProcessed, ok, failed, error}`.
 */
import { writeFile, rename, mkdir, open } from "node:fs/promises";
import { dirname } from "node:path";

const DEFAULT_RETRY_DELAYS_MS = [200, 400, 800];

function jitter(base) {
  const spread = base * 0.25;
  return base + (Math.random() * 2 - 1) * spread;
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

/**
 * Fetch with exponential backoff + jitter. Retries on network errors and
 * on HTTP 429/5xx. On final failure returns the last Response (so callers
 * can read body for diagnostics) or throws if no Response was ever obtained.
 */
export async function fetchWithRetry(url, init = {}, opts = {}) {
  const delays = opts.delaysMs ?? DEFAULT_RETRY_DELAYS_MS;
  const attempts = delays.length + 1;
  let lastErr;
  let lastResp;
  for (let i = 0; i < attempts; i++) {
    try {
      const resp = await fetch(url, init);
      lastResp = resp;
      if (resp.ok) return resp;
      if (resp.status !== 429 && resp.status < 500) return resp; // 4xx not worth retrying
    } catch (err) {
      lastErr = err;
    }
    if (i < delays.length) await sleep(jitter(delays[i]));
  }
  if (lastResp) return lastResp;
  throw lastErr ?? new Error(`fetchWithRetry failed: ${url}`);
}

/** Token-bucket rate limiter, `rps` requests per second. */
export class RateLimiter {
  constructor(rps) {
    this.intervalMs = 1000 / Math.max(0.1, rps);
    this.next = 0;
  }
  async take() {
    const now = Date.now();
    const wait = Math.max(0, this.next - now);
    this.next = Math.max(now, this.next) + this.intervalMs;
    if (wait > 0) await sleep(wait);
  }
}

/** Atomic JSON write: tmp → fsync → rename. */
export async function writeJsonAtomic(path, data) {
  await mkdir(dirname(path), { recursive: true });
  const tmp = `${path}.tmp`;
  const body = JSON.stringify(data, null, 2) + "\n";
  await writeFile(tmp, body, "utf8");
  // fsync to flush to disk
  const fh = await open(tmp, "r+");
  try {
    await fh.sync();
  } finally {
    await fh.close();
  }
  await rename(tmp, path);
}

/** Structured JSON log line to stdout. */
export function jsonLog(payload) {
  const line = {
    ts: new Date().toISOString(),
    ...payload,
    ...(payload.error
      ? { error: payload.error instanceof Error ? payload.error.message : String(payload.error) }
      : {}),
  };
  process.stdout.write(JSON.stringify(line) + "\n");
}

/** Helper for timing an operation and emitting one log line at the end. */
export async function timed(operation, fn) {
  const start = Date.now();
  try {
    const result = await fn();
    jsonLog({
      operation,
      durationMs: Date.now() - start,
      recordsProcessed: result?.recordsProcessed ?? 0,
      ok: result?.ok ?? 0,
      failed: result?.failed ?? 0,
    });
    return result;
  } catch (error) {
    jsonLog({ operation, durationMs: Date.now() - start, error });
    throw error;
  }
}
